demos/fluid/fluid.h

```diff
@@ -0,0 +1,490 @@
+          queue,
+          [&](cl::sycl::handler& cgh, auto px_a) {
+            SetBoundaryConditions(1, px_a, size, cgh);
+          },
+          px_b);
+      Submit(
+          queue,
+          [&](cl::sycl::handler& cgh, auto py_a) {
+            SetBoundaryConditions(2, py_a, size, cgh);
+          },
+          py_b);
+    }
+
+    // Project and advect the fluid velocities.
+    Project(px_b, py_b, x_b, y_b);
+    Advect(1, x_b, px_b, px_b, py_b);
+    Advect(2, y_b, py_b, px_b, py_b);
+    Project(x_b, y_b, px_b, py_b);
+
+    // Diffuse the fluid densities.
+    for (std::size_t iteration{0}; iteration < density_iterations;
+         ++iteration) {
+      Submit(
+          queue,
+          [&](cl::sycl::handler& cgh, auto previous_density_a, auto density_a) {
+            LinearSolve(0, previous_density_a, density_a, a_density,
+                        c_reciprocal_density, size, cgh);
+          },
+          previous_density_b, density_b);
+      Submit(
+          queue,
+          [&](cl::sycl::handler& cgh, auto previous_density_a) {
+            SetBoundaryConditions(0, previous_density_a, size, cgh);
+          },
+          previous_density_b);
+    }
+
+    // Advect the fluid densities.
+    Advect(0, density_b, previous_density_b, x_b, y_b);
+
+    // Update the image pixel data with the appropriate color for a given
+    // density.
+    queue.submit([&](cl::sycl::handler& cgh) {
+      auto img_acc{
+          img.template get_access<cl::sycl::access::mode::discard_write>(cgh)};
+      auto density_a{
+          density_b.template get_access<cl::sycl::access::mode::read>(cgh)};
+      cgh.parallel_for<image_kernal>(
+          cl::sycl::range<1>(size * size), [=](cl::sycl::item<1> item) {
+            auto index{item.get_id(0)};
+            auto value{density_a[index]};
+            std::uint8_t red =
+                value >= 255 ? 255 : static_cast<std::uint8_t>(value);
+            img_acc[index] = {red, 0, 0, 255};
+          });
+    });
+
+    queue.wait();
+  }
+
+  // Some aliases to improve readability of code.
+  using float_buffer = cl::sycl::buffer<float, 1>;
+  using read_write_accessor =
+      cl::sycl::accessor<float, 1, cl::sycl::access::mode::read_write,
+                         cl::sycl::access::target::global_buffer,
+                         cl::sycl::access::placeholder::false_t>;
+
+  // Wrapper around queue submission.
+  template <typename T, typename... Ts>
+  static void Submit(cl::sycl::queue& queue, T lambda, Ts&... buffers) {
+    queue.submit([&](cl::sycl::handler& cgh) {
+      lambda(cgh, CreateAccessor(cgh, buffers)...);
+    });
+  }
+
+  // Creates read_write accessors from a buffer.
+  template <typename T>
+  static read_write_accessor CreateAccessor(cl::sycl::handler& cgh, T buffer) {
+    return buffer.template get_access<cl::sycl::access::mode::read_write>(cgh);
+  }
+
+  // Clamp value to a range.
+  template <typename T>
+  static T Clamp(T value, T low, T high) {
```

Comment:
There's a SYCL builtin that can do the clamping

Reply:
Done